### PR TITLE
【No.6】feat(rollout): add full-response repetition diagnostics

### PR DESCRIPTION
**Pull request by oliver-0424**
## What

Detect repetition throughout long rollout responses and add an offline JSON diagnostic report for existing rollout dumps.

- Scan every overlapping response window, including the final unaligned window, and return suspicious character intervals, their compression ratios, the maximum ratio, and the number of scanned windows.
- Keep `has_repetition(text) -> bool` at its existing import path so the existing `repetition_frac` metric uses full-response detection.
- Add `python -m relax.entrypoints.diagnose_repetition` for training/evaluation JSONL results and trusted `.pt` dumps, with sample identifiers and JSON output.
- Add English and Chinese usage guides, regression/integration tests, and a reproducible CPU benchmark.

## Why

Previously, detection only examined the last 10,000 characters of responses longer than 10,000 characters. A response with a repeated middle section and a normal ending was therefore missed. The fixed integration example now reports the middle interval `[10000, 20000)` and a `repetition_frac` of `0.5` across the repetitive sample and its nonrepetitive control.

**Metric comparability:** The previous metric always returned false for responses of 10,000 characters or fewer. The new detector checks those responses as well as repetition anywhere in longer responses, so `repetition_frac` from old and new runs should not be compared directly.

Relates to #321, contributor program cohort 2, **No.6: 长响应全文重复检测与离线诊断**.

- [Task requirements](https://github.com/redai-studio/community/blob/main/contributor-program/2026-cohort-2/official-task.md#no6长响应全文重复检测与离线诊断)
- Claim: https://github.com/redai-studio/Relax/issues/321#issuecomment-5724946727

## How

The detector uses 10,000-character windows, a 5,000-character stride, and the existing default compression criterion: UTF-8 byte length divided by zlib level-9 compressed length, strictly greater than `10.0`. With the default `stop_after_first_hit=False`, it scans every window and adds the final window when the response length is not aligned to the stride. The existing Boolean `has_repetition` wrapper passes `stop_after_first_hit=True` and returns after the first hit. When early exit is enabled, result fields describe only the scanned windows; the offline CLI keeps the exhaustive default. Window size, stride, compression and threshold are identical in both modes.

Offsets are zero-based Unicode code-point indices with an exclusive end. A hit identifies a suspicious window, not the exact repetition boundary. Empty responses produce no windows and a maximum ratio of `0.0`. The detector operates on `Sample.response`, including any tool observations stored in that field, without changing rewards or generation.

The existing `compression_ratio` helper is unchanged. A lightweight detector module keeps the JSONL diagnostic path independent of training dependencies; PyTorch is imported only when reading a trusted `.pt` file. JSONL records are read incrementally, while the report retains per-sample results without copying prompt or response text.

## Testing

```bash
pre-commit run --all-files --show-diff-on-failure
python -m pytest \
  tests/utils/test_repetition.py \
  tests/utils/test_repetition_integration.py \
  tests/utils/test_metric_utils_rloo.py \
  tests/utils/test_multimodal_rollout_stats.py -q -rs
```

Result: **62 passed, 1 skipped**. The existing Megatron-specific test is skipped because `megatron.core` is unavailable; all new No.6 tests execute and pass.

- Detector coverage: beginning/middle/end repetition, a normal tail after middle repetition, empty/short/exact-window responses, overlapping boundaries, an unaligned final window, Chinese/emoji/combining-character offsets, nonrepetitive controls, equality at the threshold, custom parameters, and all 199 windows of a 1,000,000-character response.
- Early-exit coverage: real compression-call counts for beginning/middle/end hits and nonrepetitive controls, partial result fields, strict default-threshold equality, both modes rejecting invalid options on empty input, and complete diagnostics retaining a later, higher compression ratio.
- Integration coverage: the complete real rollout module's `compute_metrics_from_samples`, real training/evaluation `.pt` and JSONL writers, and the diagnostic CLI in subprocesses. Only deployment-specific dependencies are stubbed; the metric function, samples, detector, dump serialization, and readers remain real.
- CLI coverage: JSON stdout/file output, original sample IDs, empty inputs, malformed records, and preventing the output path from replacing the input.
- Both language guides, command examples, report fields, sidebar entries, and relative links were checked.

- [x] `pre-commit run --all-files` passes on the current commit
- [x] Targeted CPU tests pass
- [ ] Full `pytest tests/` / multi-node GPU validation — not run locally because the full training dependencies and a configured multi-node GPU cluster are unavailable
- [x] New tests added
- [x] Documentation updated

## Type of Change

- [x] Bug fix
- [x] New feature
- [ ] Breaking change
- [x] Documentation update
- [ ] Refactoring
- [x] Performance improvement
- [ ] CI/CD or build changes

### CPU benchmark

```bash
PYTHONPATH=. python scripts/benchmarks/benchmark_repetition.py --runs 7 > benchmark.json
```

Environment: Python 3.12.13, zlib 1.3.1, Linux x86-64 / WSL2, seed 42. Each case runs in a fresh process, with a warmup and seven timed runs. The following table measures the current exhaustive diagnostic mode on one sample. Traced peak measures one full diagnostic call in a separate untimed run; process peak RSS includes interpreter imports and input construction.

| Characters | Pattern | Windows | Hits | Median CPU (ms) | Median wall (ms) | Traced peak (KiB) | Process peak RSS (MiB) |
|---:|---|---:|---:|---:|---:|---:|---:|
| 10,000 | control | 1 | 0 | 0.159 | 0.158 | 303.9 | 18.9 |
| 10,000 | middle | 1 | 1 | 0.062 | 0.062 | 303.9 | 18.8 |
| 10,000 | repetitive | 1 | 1 | 0.065 | 0.064 | 303.9 | 18.8 |
| 100,000 | control | 19 | 0 | 3.395 | 3.394 | 304.0 | 18.7 |
| 100,000 | middle | 19 | 1 | 3.072 | 3.071 | 304.2 | 18.7 |
| 100,000 | repetitive | 19 | 19 | 1.091 | 1.090 | 306.9 | 18.9 |
| 1,000,000 | control | 199 | 0 | 35.017 | 35.014 | 304.0 | 26.6 |
| 1,000,000 | middle | 199 | 1 | 34.940 | 34.938 | 304.2 | 26.6 |
| 1,000,000 | repetitive | 199 | 199 | 11.601 | 11.600 | 337.6 | 20.0 |

`control` is seeded ASCII letter/digit text; `middle` replaces its middle 10,000 characters with `a`; `repetitive` is entirely `a`. At 10,000 characters, the latter two inputs are identical.

### Batch comparison: exhaustive and early-exit modes

```bash
PYTHONPATH=. python scripts/benchmarks/benchmark_repetition.py --runs 5 --batch-size 256 --lengths 100000 500000 --patterns control beginning middle end repetitive > batch.json
```

Each batch repeats the same input 256 times; values are median CPU milliseconds over five runs. Both columns use the current detector. The first column collects complete diagnostics; the second enables `stop_after_first_hit=True`, as used by the Boolean metric. This measures detector cost, with input construction outside the timer.

| Characters/sample | Pattern | Full scan / batch (ms) | Early exit / batch (ms) |
|---:|---|---:|---:|
| 100,000 | control | 860.466 | 857.206 |
| 100,000 | beginning | 808.588 | 12.396 |
| 100,000 | middle | 789.117 | 397.965 |
| 100,000 | end | 807.330 | 807.219 |
| 100,000 | repetitive | 282.242 | 15.085 |
| 500,000 | control | 4482.183 | 4476.509 |
| 500,000 | beginning | 4426.908 | 12.486 |
| 500,000 | middle | 4410.601 | 2213.835 |
| 500,000 | end | 4489.019 | 4458.329 |
| 500,000 | repetitive | 1479.303 | 15.210 |

`beginning`, `middle` and `end` insert one 10,000-character repeated region at the named position in the same control input. Early or middle hits reduce work; a last-window hit or no hit still requires scanning all windows. These timings do not imply the same speedup for an entire training step.

### Ablation validation

The exhaustive diagnostic contract was checked using isolated mutations of the initial detector at `fed51082`, against the same 37 No.6 tests. The current early-exit option is opt-in for the Boolean metric; default diagnostics still require every window and all hits.

| Isolated variant | Failing tests |
|---|---:|
| Submitted full-response implementation | 0 (37 passed) |
| Replace full scan with tail-only detection | 20 |
| Remove overlap | 8 |
| Omit the unaligned final window | 4 |
| Change strict `>` threshold to `>=` | 1 |
| Prematurely stop exhaustive diagnostics after the first hit | 8 |

The fixed integration samples use `random.Random(42)` and include a 30,000-character response with repetition only at `[10000, 20000)`, plus a nonrepetitive control. Both tails have compression ratios below 10; the real metric and offline report still detect exactly one repetitive sample. This directly covers the failure mode in the original tail-only implementation.

### CODE REVIEW

**Comment by oliver-0424** [comment]
Hi @SigureMo, the CI and GPU Unit Tests workflows for this PR are currently awaiting maintainer approval (`action_required`), possibly due to the approval requirement for first-time contributors 😭

When you have a moment, could you please click **Approve workflows to run** so the checks can proceed? Thanks for your help!

**Comment by oliver-0424** [comment]
@rai-studio-bot Please review this PR.

File: relax/utils/metrics/metric_utils.py [hunk 1]
```diff
@@ -269,11 +272,8 @@ def compression_ratio(
     return ratio, savings_pct
 
 
-def has_repetition(text: str):
-    if len(text) > 10000 and compression_ratio(text[-10000:])[0] > 10:
-        return True
-    else:
-        return False
+def has_repetition(text: str) -> bool:
+    return detect_repetition(text).has_repetition
```
[the remark this answers is not in the document]
Comment (oliver-0424) [reply relax/utils/metrics/metric_utils.py]: @rai-studio-bot Addressed in `ea22ddf`.

Added `stop_after_first_hit=False` to `detect_repetition`. The Boolean metric wrapper enables it, while offline diagnostics still scan all windows. Early-exit result fields describe only the scanned windows.

Local validation: **62 passed, 1 skipped** (the existing Megatron test); full pre-commit passes. For 256 samples of 100k characters each, full-scan vs early-exit median CPU time is **808.6 vs 12.4 ms** for beginning hits and **789.1 vs 398.0 ms** for middle hits. No-hit and last-window cases remain essentially unchanged.

The PR description and both language guides now explain the changed `repetition_frac` semantics and limits on comparing historical runs.

Please review the update.

**Comment by oliver-0424** [comment]
Hi @SigureMo @yuanlehome, task No.6 is complete and ready for human review. All five CI checks are passing, and the review bot has verified the latest commit and confirmed that its feedback is resolved. Could you please take a look when you have a moment? Thank you!

**Comment by SigureMo** [comment]
辛苦 merge 下最新 main，新增了两条流水线

**Review by yuanlehome — APPROVED** [review]
